src/main/java/persistence/core/EntityMetadata.java

```diff
@@ -5,20 +5,25 @@
 import persistence.exception.PersistenceException;
 
 import java.util.List;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.stream.Collectors;
 
 public class EntityMetadata<T> {
 
+    private final Class<T> clazz;
     private final String tableName;
     private final EntityColumns columns;
-    private final EntityColumn idColumn;
+    private final EntityIdColumn idColumn;
+    private final List<EntityOneToManyColumn> oneToManyColumns;
 
     public EntityMetadata(final Class<T> clazz) {
         this.validate(clazz);
+        this.clazz = clazz;
         this.tableName = initTableName(clazz);
         this.columns = new EntityColumns(clazz);
         this.idColumn = this.columns.getId();
+        this.oneToManyColumns = this.columns.getOneToManyColumns();
```

Comment:
```suggestion
    public EntityMetadata(final Class<T> clazz) {
        this.validate(clazz);
        this.clazz = clazz;
        this.tableName = initTableName(clazz);
        this.columns = new EntityColumns(clazz);
   
        ...     
        
    public String getIdColumnName() {
        return this.columns.idName();
    }

    public List<EntityOneToManyColumn> getOneToManyColumns() {
        return this.columns.oneTomanyColumns();
    }
```

궁금한게 있어요!
`this.idColumn`, `this.oneToManyColumns` 의 내부 상태를 가져가는 이유가 있을까요?
`EntityColumns` 일급콜렉션을 잘 만들어주셨다고 생각해요~
그렇다면 `EntityColumns` 에게 그때 그때 물어보는 관점은 어떻게 생각하시나요~? 🤔

Reply:
이 부분은 제가 개발할때 가지고 있는 딜레마 중 하나가 표출된 부분입니다. 
불변객체 와 일급컬렉션의 조합을 통해 해당 컬렉션은 한번 만들어 지고 난 뒤, 단순 조회성 메서드의 반환값은 항상 같습니다. 그렇다보니 딜레마가 생기는데요, 메서드 실행시 이미 계산된걸 반환하냐, 아니면 항상 같은값을 반환하지만 계산하냐 입니다. 그 중 Id 와 OneToManyColumns 는 좀 특이케이스라 생각해서 필드로 가지고 싶더라구요.

말씀하신대로 EntityColumns 에게 그때그때 물어본다면 마찬가지로 필드로 가지고 있을지, 계산할지 똑같이 고민할거같은데 제가 아는 일급컬렉션의 조건 중 하나가 `컬렉션을 제외한 다른 멤버 변수가 없어야 한다` 입니다. 그럼 필드로 가지지 말고 항상 계산하는 방식으로 만들면 되는걸까요? 아니면 아이덴티컬한 필드는 따로 가지고 있는게 나을까요?

Reply:
@jongmin4943 
그렇네요. 확실히 성능적으로 보면 필드로 가지고 똑같이 반환하는게 이점이 있어보이네요! 🤔 
`private final List<EntityOneToManyColumn> oneToManyColumns;` 
저도 고민해보았는데, 지금 구조상 구현하신것처럼 필드로 가지는것도 좋을것 같네요 ㅎ!

사실 저의 핵심은, 왠만하면 객체에게 메시지를 던지고 물어보자 였어요~ 🙏 
그러한 관점으로 아래처럼 리팩토링 해보았어요.
```
# DdlGenerator
...
private String generateAssociatedColumnsClause(final EntityMetadata<?> entityMetadata) {  
	...
	final Set<EntityMetadata<?>> allAssociatedEntitiesMetadata = EntityMetadataProvider.getInstance().getAllAssociatedEntitiesMetadata(entityMetadata);

```


`getAllAssociatedEntitiesMetadata` 메소드를 객체에게 물어보는 관점으로 리팩토링 해볼께요!

1. `getAllAssociatedEntitiesMetadata` 호출합니다

2. `entityMetadataCache(EntityMetadata)` 에게 `hasAssociatedOf` 물어봅니다. 
```
# EntityMetadataProvider
public Set<EntityMetadata<?>> getAllAssociatedEntitiesMetadata(final EntityMetadata<?> entityMetadata) {  
	return cache.values().stream()  
		.filter(entityMetadataCache -> entityMetadataCache.hasAssociatedOf(entityMetadata))  
		.collect(Collectors.toUnmodifiableSet());  
}

```

3. `hasAssociatedOf`
```
# EntityMetadata
public class EntityMetadata<T> {
	...

	public boolean hasAssociatedOf(EntityMetadata<?> entityMetadata) {  
		return columns.containType(entityMetadata);  
	}
```

4. `columns` 을 가진 `entityMetadata` 에게 검증하라고 시킨다.
```
# EntityColumns
public class EntityColumns implements Iterable<EntityColumn> {
    private final List<EntityColumn> columns;
    private final List<EntityOneToManyColumn> oneToManyColumns;

    public EntityColumns(final Class<?> clazz) {
        this.columns = generateColumns(clazz);
        this.oneToManyColumns = getOneToManyColumns();
    }
...

	public boolean containType(EntityMetadata<?> entityMetadata) {  
		return oneToManyColumns.stream()  
			.anyMatch(entityOneToManyColumn -> entityMetadata.isType(entityOneToManyColumn.getJoinColumnType()));  
	}
```

분명 비효율적? 아니면 너무 과한거 아닌가? 로 보일수 있는점 이라면 충분히 공감됩니다! 
하지만, 이렇게 하다보면 
적절한 상태의 위치에 대한 고민, 메시지를 통한 적절한 책임인가에 대한 고민등 **확장된 사고**, **새로운 길**이 보이지 않을까요? 😄 
지금은 실전이 아니라 훈련중이니까요!

> 리팩토링 해보았는데 테스트 코드 정말 잘 구현해놓으셔서 **맘놓고** 리팩토링 할 수 있었네요 💯 💯 💯

Reply:
반영완료 - [`7d49f97`](https://github.com/next-step/jpa-association/pull/18/commits/7d49f97c70995e560e82f8f85be8e6faac3f55b3)
이 부분은 `EntityColumns` 까지 가지 않고  `EntityMetadata` 에서 처리했는데요, `EntityColumns` 가 `EntityMetadata` 라는 객체를 알아야하나? 라는 고민이 들더라구요.